WebDriverAgentLib/Utilities/FBClassChainQueryParser.m

```diff
@@ -0,0 +1,371 @@
+  return @[];
+}
+
++ (BOOL) canConsumeCharacter:(unichar)character
+{
+  return [self.allowedCharacters characterIsMember:character];
+}
+
+- (instancetype)nextTokenWithCharacter:(unichar)character
+{
+  if ([self.class canConsumeCharacter:character] && self.asString.length < [self.class maxLength]) {
+    NSMutableString *value = [NSMutableString stringWithString:self.asString];
+    [value appendFormat:@"%C", character];
+    self.asString = value.copy;;
+    return self;
+  }
+  for (Class matchingTokenClass in self.followingTokens) {
+    if ([matchingTokenClass canConsumeCharacter:character]) {
+      return [[[matchingTokenClass alloc] init] nextTokenWithCharacter:character];
+    }
+  }
+  return nil;
+}
+
+@end
+
+
+@implementation ClassNameToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet letterCharacterSet];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[SplitterToken.class, OpeningBracketToken.class];
+}
+
+@end
+
+
+@implementation StarToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"*"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[SplitterToken.class, OpeningBracketToken.class];
+}
+
++ (unsigned int)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation SplitterToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"/"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[StarToken.class, ClassNameToken.class];
+}
+
++ (unsigned int)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation OpeningBracketToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"["];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[NumberToken.class];
+}
+
++ (unsigned int)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation NumberToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  NSMutableCharacterSet *result = [NSMutableCharacterSet new];
+  [result formUnionWithCharacterSet:[NSCharacterSet decimalDigitCharacterSet]];
+  [result addCharactersInString:@"-"];
+  return result.copy;
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[ClosingBracketToken.class];
+}
+
+@end
+
+
+@implementation ClosingBracketToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"]"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[SplitterToken.class];
+}
+
++ (unsigned int)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation FBClassChainElement
+
+- (id) initWithType:(XCUIElementType)type position:(int)position
+{
+  self = [super init];
+  if (self) {
+    _type = type;
+    _position = position;
+  }
+  return self;
+}
+
+@end
+
+
+@implementation FBClassChainQueryParser
+
++ (void)tokenizationErrorWithIndex:(unsigned int)index originalQuery:(NSString *)originalQuery error:(NSError **)error
```

Comment:
If you accept an error you should return a `BOOL` to indicate whether operation was successful.